File: backend/village-profile/app/models/vp/household.py
from django.contrib.auth.models import User
from django.db import models


class Household(models.Model):

    RESIDENT_CHOICES = (
        ('जन्मसिद्ध', 'from_birth'),
        ('बसाईसराई', 'migrated'),
    )

    id_string = models.TextField(max_length=500, null=True)

    hoh_name = models.CharField(max_length=50, null=True)
    hoh = models.ForeignKey('Member', on_delete=models.DO_NOTHING, related_name='dead', null=True)
    province = models.ForeignKey('Province', on_delete=models.DO_NOTHING, null=True)
    district = models.ForeignKey('District', on_delete=models.DO_NOTHING, null=True)
    local_level = models.ForeignKey('LocalLevel', on_delete=models.DO_NOTHING, null=True)
    ward = models.ForeignKey('Ward', on_delete=models.DO_NOTHING, null=True)
    basti = models.ForeignKey('Basti', on_delete=models.DO_NOTHING, null=True)
    marga = models.ForeignKey('Marga', on_delete=models.DO_NOTHING, null=True)
    religion = models.ForeignKey('Religion', on_delete=models.DO_NOTHING, null=True)
    jaati = models.ForeignKey('Jaati', on_delete=models.DO_NOTHING, null=True)
    mother_tongue = models.ForeignKey('MotherTongue', on_delete=models.DO_NOTHING, null=True)
    main_occupation = models.ForeignKey('MainOccupation', on_delete=models.DO_NOTHING, null=True)

    has_bank_acc = models.IntegerField(default=0)
    has_cooperative_acc = models.IntegerField(default=0)
    has_garden = models.IntegerField(default=0)
    member_with_life_insurance = models.IntegerField(null=True)
    member_with_health_insurance = models.IntegerField(null=True)
    responder_name = models.CharField(max_length=200, null=True)
    house_num = models.IntegerField(null=True)
    num_of_member = models.IntegerField(null=True)
    resident_type = models.CharField(max_length=14, choices=RESIDENT_CHOICES, default='जन्मसिद्ध')
    phone_num = models.BigIntegerField(null=True)
    mobile_num = models.BigIntegerField(null=True)
    longitude = models.DecimalField(decimal_places=5, max_digits=12, null=True)
    latitude = models.DecimalField(decimal_places=5, max_digits=12, null=True)
    geo_location = models.CharField(max_length=500, null=True)
    altitude = models.DecimalField(decimal_places=5, max_digits=12, null=True)
    accuracy = models.DecimalField(decimal_places=5, max_digits=12, null=True)
    responder_image = models.ImageField(upload_to='image', null=True)

    step = models.IntegerField(default=1)
    status = models.IntegerField(null=True)
    remarks = models.TextField(null=True)
    office = models.ForeignKey('Office', on_delete=models.CASCADE, null=True)
    user = models.ForeignKey(User, on_delete=models.DO_NOTHING, null=True)
    created_at = models.DateTimeField(auto_now_add=True, null=True)
    updated_at = models.DateTimeField(auto_now=True, null=True)

    def __str__(self):
        return self.hoh_name

File: app/src/db/models/Household.ts
import { db } from "../db";

export interface IHousehold {
  id?: number;
  id_string?: String;
  hoh_name?: String;
  hoh?: String;
  province?: String;
  district?: String;
  local_level?: String;
  ward_id?: String;
  basti_id?: String;
  marga_id?: String;
  religion?: String;
  jaati?: String;
  mother_tongue?: String;
  main_occupation?: String;
  has_bank_acc?: String;
  has_cooperative_acc?: String;
  has_garden?: String;
  member_with_life_insurance?: String;
  member_with_health_insurance?: String;
  responder_name?: String;
  house_num?: String;
  num_of_member?: String;
  resident_type?: String;
  phone_num?: String;
  mobile_num?: String;
  longitude?: String;
  latitude?: String;
  geo_location?: String;
  altitude?: String;
  accuracy?: String;
  responder_image?: String;
  step?: String;
  status?: String;
  remarks?: String;
  office?: String;
  user_id?: String;
  is_posted?: String;
}
export class Household {
  id: number;
  hoh_name?: String;
  hoh?: String;
  ward_id?: String;
  basti_id?: String;
  marga_id?: String;
  religion?: String;
  jaati?: String;
  mother_tongue?: String;
  main_occupation?: String;
  has_bank_acc?: String;
  has_cooperative_acc?: String;
  has_garden?: String;
  member_with_life_insurance?: String;
  member_with_health_insurance?: String;
  responder_name?: String;
  house_num?: String;
  num_of_member?: String;
  resident_type?: String;
  phone_num?: String;
  mobile_num?: String;
  longitude?: String;
  latitude?: String;
  geo_location?: String;
  altitude?: String;
  accuracy?: String;
  responder_image?: String;
  step?: String;
  status?: String;
  remarks?: String;
  user_id?: String;
  is_posted?: String;

  constructor(data: IHousehold) {
    this.hoh_name = data.hoh_name;
    this.hoh = data.hoh;
    this.ward_id = data.ward_id;
    this.basti_id = data.basti_id;
    this.marga_id = data.marga_id;
    this.religion = data.religion;
    this.jaati = data.jaati;
    this.mother_tongue = data.mother_tongue;
    this.main_occupation = data.main_occupation;
    this.has_bank_acc = data.has_bank_acc;
    this.has_cooperative_acc = data.has_cooperative_acc;
    this.has_garden = data.has_garden;
    this.member_with_life_insurance = data.member_with_life_insurance;
    this.member_with_health_insurance = data.member_with_health_insurance;
    this.responder_name = data.responder_name;
    this.house_num = data.house_num;
    this.num_of_member = data.num_of_member;
    this.resident_type = data.resident_type;
    this.phone_num = data.phone_num;
    this.mobile_num = data.mobile_num;
    this.longitude = data.longitude;
    this.latitude = data.latitude;
    this.geo_location = data.geo_location;
    this.altitude = data.altitude;
    this.accuracy = data.accuracy;
    this.responder_image = data.responder_image;
    this.step = data.step;
    this.status = data.status;
    this.remarks = data.remarks;
    this.user_id = data.user_id;
    this.is_posted = data.is_posted;
    db.households.mapToClass(Household);
  }
  save() {
    return db.households.put(this);
  }
}

export async function addNewHousehold(data: IHousehold) {
  await db.transaction("rw", db.households, async function () {
    await db.households.add(
      new Household({...data})
    );
  });
}

export async function getAllHousehold() {
  return await db.transaction("r", db.households, async function () {
    let households = await db.households.toArray();
    return households;
  });
}

export async function getHouseholdById(id: number) {
  return await db.households.get(id);
}

export async function getPendingHouseholds(user_id: string) {
  return await db.households
    .where("[user_id+is_posted]").equals([user_id,"0"]).toArray();
}

export async function updateHousehold(data: IHousehold) {
  return await db.households.put({...data});
}
